In [ ]:
%%bash
cd ..
source ./activateEnv.sh


In [2]:
%%bash
# Instalar os software que vamos usar nesta prática
mamba install -c bioconda -c conda-forge fastqc bwa gatk4 cutadapt samtools bedtools ensembl-vep=115 -y


Pinned packages:

  - python=3.10



error    libmamba Could not create directory '/usr/pkgs': Permission denied



Transaction

  Prefix: /home/figa/Documents/Programming/VSCode/Metagenomics-Course/env

  All requested packages already installed


Transaction starting

Transaction finished



In [ ]:
%%bash

mkdir -p fastq bwa bed cutadapt raw reference gatk vep 
mv *.fastq.gz raw/
gunzip -c Chr7-reference.fasta.gz > reference/Chr7-reference.fasta
rm Chr7-reference.fasta.gz


In [4]:
%%bash 
fastqc -o fastq/ raw/caso3-tumor-pulmonar_R1.fastq.gz raw/caso3-tumor-pulmonar_R2.fastq.gz
cutadapt \
  -u 5 -U 5 \
  -u -5 -U -5 \
  -m 90 \
  -o cutadapt/caso3-tumor-pulmonar-trimmed_R1.fastq.gz \
  -p cutadapt/caso3-tumor-pulmonar-trimmed_R2.fastq.gz \
  raw/caso3-tumor-pulmonar_R1.fastq.gz \
  raw/caso3-tumor-pulmonar_R2.fastq.gz


application/gzip
application/gzip


Started analysis of caso3-tumor-pulmonar_R1.fastq.gz
Approx 5% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 10% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 15% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 20% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 25% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 30% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 35% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 40% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 45% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 50% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 55% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 60% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 65% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 70% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 75% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 80% complete for caso3-tumor-pulmonar_R1.fastq.gz
Approx 85% complete for caso3-tumor-

Analysis complete for caso3-tumor-pulmonar_R1.fastq.gz


Started analysis of caso3-tumor-pulmonar_R2.fastq.gz
Approx 5% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 10% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 15% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 20% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 25% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 30% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 35% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 40% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 45% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 50% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 55% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 60% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 65% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 70% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 75% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 80% complete for caso3-tumor-pulmonar_R2.fastq.gz
Approx 85% complete for caso3-tumor-

Analysis complete for caso3-tumor-pulmonar_R2.fastq.gz
This is cutadapt 5.1 with Python 3.10.14
Command line parameters: -u 5 -U 5 -u -5 -U -5 -m 90 -o cutadapt/caso3-tumor-pulmonar-trimmed_R1.fastq.gz -p cutadapt/caso3-tumor-pulmonar-trimmed_R2.fastq.gz raw/caso3-tumor-pulmonar_R1.fastq.gz raw/caso3-tumor-pulmonar_R2.fastq.gz
Processing paired-end reads on 1 core ...

=== Summary ===

Total read pairs processed:          4,000,000

== Read fate breakdown ==
Pairs that were too short:                   1 (0.0%)
Pairs written (passing filters):     3,999,999 (100.0%)

Total basepairs processed: 1,199,968,732 bp
  Read 1:   599,968,852 bp
  Read 2:   599,999,880 bp
Total written (filtered):  1,119,968,504 bp (93.3%)
  Read 1:   559,968,763 bp
  Read 2:   559,999,741 bp


In [7]:
%%bash
#1o indexa contra o genoma humano pra depois alinhar
samtools faidx reference/Chr7-reference.fasta
bwa index reference/Chr7-reference.fasta

bwa mem reference/Chr7-reference.fasta \
  -R '@RG\tID:sample1\tSM:sample1\tPL:ILLUMINA\tLB:lib1\tPU:unit1' \
  cutadapt/caso3-tumor-pulmonar-trimmed_R1.fastq.gz \
  cutadapt/caso3-tumor-pulmonar-trimmed_R2.fastq.gz \
  | samtools sort -o bwa/caso3-tumor-pulmonar-mapped-host.bam
  
  samtools index bwa/caso3-tumor-pulmonar-mapped-host.bam

[bwa_index] Pack FASTA... 0.31 sec
[bwa_index] Construct BWT for the packed sequence...
[BWTIncCreate] textLength=318691946, availableWord=34423912
[BWTIncConstructFromPacked] 10 iterations done. 56784042 characters processed.
[BWTIncConstructFromPacked] 20 iterations done. 104904522 characters processed.
[BWTIncConstructFromPacked] 30 iterations done. 147670090 characters processed.
[BWTIncConstructFromPacked] 40 iterations done. 185676202 characters processed.
[BWTIncConstructFromPacked] 50 iterations done. 219452122 characters processed.
[BWTIncConstructFromPacked] 60 iterations done. 249468234 characters processed.
[BWTIncConstructFromPacked] 70 iterations done. 276142554 characters processed.
[BWTIncConstructFromPacked] 80 iterations done. 299846746 characters processed.
[bwt_gen] Finished constructing BWT in 89 iterations.
[bwa_index] 45.71 seconds elapse.
[bwa_index] Update BWT... 0.32 sec
[bwa_index] Pack forward-only FASTA... 0.17 sec
[bwa_index] Construct SA from BWT and Occ.

In [3]:
%%bash
# Indexar a referência e BAM para chamada de variantes com GATK
samtools dict reference/Chr7-reference.fasta > reference/Chr7-reference.dict
samtools index bwa/caso3-tumor-pulmonar-mapped-host.bam

 gatk Mutect2 \
 -R reference/Chr7-reference.fasta \
    -I bwa/caso3-tumor-pulmonar-mapped-host.bam \
    -O mutect/caso3-tumor-pulmonar-unfiletered.vcf

Using GATK jar /home/figa/Documents/Programming/VSCode/Metagenomics-Course/env/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /home/figa/Documents/Programming/VSCode/Metagenomics-Course/env/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar Mutect2 -R reference/Chr7-reference.fasta -I bwa/caso3-tumor-pulmonar-mapped-host.bam -O mutect/caso3-tumor-pulmonar-unfiletered.vcf
09:36:50.265 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/figa/Documents/Programming/VSCode/Metagenomics-Course/env/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
09:36:50.313 INFO  Mutect2 - ------------------------------------------------------------
09:36:50.314 INFO  Mutect2 - The Genome Analysis Toolkit (GATK) v4.6.2.0
09:36:50.314 INFO  Mutect2 - Fo

Tool returned:
SUCCESS


# Anotação de Variantes

In [14]:
%%bash
vep_install -a cf -s homo_sapiens -y GRCh38 --NO_UPDATE ##baixa o genoma humano
##Oh fuck.

 - getting list of available cache files
 - downloading https://ftp.ensembl.org/pub/release-115/variation/indexed_vep_cache/homo_sapiens_vep_115_GRCh38.tar.gz
 - unpacking homo_sapiens_vep_115_GRCh38.tar.gz
 - converting cache, this may take some time but will allow VEP to look up variants and frequency data much faster
 - use CTRL-C to cancel if you do not wish to convert this cache now (you may run convert_cache.pl later)
2025-10-12 12:37:45 - Processing homo_sapiens
2025-10-12 12:37:45 - Processing version 115_GRCh38
2025-10-12 12:37:45 - No unprocessed types remaining, skipping
2025-10-12 12:37:45 - All done!
 - downloading Homo_sapiens.GRCh38.dna.toplevel.fa.gz
 - downloading Homo_sapiens.GRCh38.dna.toplevel.fa.gz.fai
 - downloading Homo_sapiens.GRCh38.dna.toplevel.fa.gz.gzi

The FASTA file should be automatically detected by the VEP when using --cache or --offline.
If it is not, use "--fasta /home/figa/.vep/homo_sapiens/115_GRCh38/Homo_sapiens.GRCh38.dna.toplevel.fa.gz"


All don

In [9]:
%%bash
bgzip gatk/caso3-tumor-pulmonar-host-variants.vcf
tabix -p vcf gatk/caso3-tumor-pulmonar-host-variants.vcf.gz
#gera os arquivos VCF para char o gene correto

In [10]:
%%bash
# Anotação de variantes a nível de proteína
vep -i gatk/caso3-tumor-pulmonar-host-variants.vcf.gz \
    --output_file vep/caso3-tumor-pulmonar-host-variants.vep.tsv \
    --tab --force_overwrite \
    --offline --cache --dir_cache ~/.vep \
    --assembly GRCh38 \
    --symbol --hgvs --canonical --af_1kg --max_af \
    --everything --pick\
    --stats_file vep/caso3-tumor-pulmonar-host-variants.vep.html

# Análise e Interpretação dos Resultados

In [7]:
import pandas as pd
import numpy as np
from io import StringIO

# Robustly load VEP TSV so the header (the line starting with a single '#') is preserved
vep_path = 'vep/caso3-tumor-pulmonar-host-variants.vep.tsv'
with open(vep_path, 'r', encoding='utf-8') as fh:
    lines = fh.readlines()

# Find the header line that starts with a single '#'<not with '##'>
header_idx = None
for i, line in enumerate(lines):
    if line.startswith('#') and not line.startswith('##'):
        header_idx = i
        break

if header_idx is None:
    # Fallback: try simple read (may fail if header was commented)
    vep_df = pd.read_csv(vep_path, sep='\t', comment='#')
else:
    header_line = lines[header_idx].lstrip('#').strip()
    headers = header_line.split('\t')
    data = ''.join(lines[header_idx+1:])
    vep_df = pd.read_csv(StringIO(data), sep='\t', names=headers, header=None)

# Summary
print(f"Total de variantes anotadas: {len(vep_df)}")
print('\nColunas (primeiras 40):')
print(list(vep_df.columns)[:40])

# Show first rows
vep_df.head(3)

Total de variantes anotadas: 2

Colunas (primeiras 40):
['Uploaded_variation', 'Location', 'Allele', 'Gene', 'Feature', 'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'IMPACT', 'DISTANCE', 'STRAND', 'FLAGS', 'VARIANT_CLASS', 'SYMBOL', 'SYMBOL_SOURCE', 'HGNC_ID', 'BIOTYPE', 'CANONICAL', 'MANE', 'MANE_SELECT', 'MANE_PLUS_CLINICAL', 'TSL', 'APPRIS', 'CCDS', 'ENSP', 'SWISSPROT', 'TREMBL', 'UNIPARC', 'UNIPROT_ISOFORM', 'GENE_PHENO', 'SIFT', 'PolyPhen', 'EXON', 'INTRON', 'DOMAINS']


Uploaded_variation                Location Allele             Gene  \
0   NC_000007.14_55191822_T/G   NC_000007.14:55191822      G  ENSG00000146648   
1  NC_000007.14_140753336_A/T  NC_000007.14:140753336      T  ENSG00000157764   

           Feature Feature_type       Consequence  cDNA_position  \
0  ENST00000275493   Transcript  missense_variant           2834   
1  ENST00000646891   Transcript  missense_variant           2025   

   CDS_position  Protein_position  ...  MAX_AF_POPS  \
0          2573               858  ...            -   
1          1799               600  ...  gnomADe_SAS   

                                    CLIN_SIG    SOMATIC      PHENO  \
0                   drug_response,pathogenic      0,1,1      1,1,1   
1  not_provided,likely_pathogenic,pathogenic  0,0,1,1,1  1,1,1,1,1   

                                              PUBMED  MOTIF_NAME MOTIF_POS  \
0  24944790,31019283,22992668,27618021,32295184,2...           -         -   
1  23757202,30089490,26900293,12460918,20619739,2...           -         -   

  HIGH_INF_POS MOTIF_SCORE_CHANGE TRANSCRIPTION_FACTORS  
0            -                  -                     -  
1            -                  -                     -  

[2 rows x 82 columns]

In [12]:
# Filtrar variantes de alto impacto (missense, frameshift, nonsense, etc.)
# Focar em variantes não-sinônimas que alteram a proteína
pattern = r'missense|frameshift|stop_gained|stop_lost|start_lost|inframe'
high_impact = vep_df[vep_df['Consequence'].str.contains(
    pattern,
    case=False,
    na=False,
    regex=True
)]

print(f"\nVariantes de alto impacto: {len(high_impact)}")
print("\nGenes afetados:")
if 'SYMBOL' in high_impact.columns:
    print(high_impact['SYMBOL'].value_counts())
    
# Exibir detalhes das variantes de alto impacto
print("\nDetalhes das variantes de alto impacto:")
columns_to_show = ['Location', 'SYMBOL', 'Consequence', 'HGVSc', 'HGVSp', 
                   'IMPACT', 'CANONICAL', 'AF', 'MAX_AF']
available_cols = [col for col in columns_to_show if col in high_impact.columns]
high_impact[available_cols]


Variantes de alto impacto: 2

Genes afetados:
SYMBOL
EGFR    1
BRAF    1
Name: count, dtype: int64

Detalhes das variantes de alto impacto:


Location SYMBOL       Consequence  \
0   NC_000007.14:55191822   EGFR  missense_variant   
1  NC_000007.14:140753336   BRAF  missense_variant   

                         HGVSc                          HGVSp    IMPACT  \
0  ENST00000275493.7:c.2573T>G  ENSP00000275493.2:p.Leu858Arg  MODERATE   
1  ENST00000646891.2:c.1799T>A  ENSP00000493543.1:p.Val600Glu  MODERATE   

  CANONICAL AF     MAX_AF  
0       YES  -          -  
1       YES  -  1.159e-05

In [10]:
# Filtrar por genes conhecidos de câncer de pulmão
cancer_genes = ['TP53', 'EGFR', 'KRAS', 'ALK', 'ROS1', 'BRAF', 'MET', 
                'RET', 'ERBB2', 'PIK3CA', 'STK11', 'KEAP1']

if 'SYMBOL' in vep_df.columns:
    cancer_variants = vep_df[vep_df['SYMBOL'].isin(cancer_genes)]
    
    if len(cancer_variants) > 0:
        print(f"Variantes encontradas em genes conhecidos de câncer de pulmão: {len(cancer_variants)}")
        print("\nGenes identificados:")
        print(cancer_variants['SYMBOL'].value_counts())
        
        print("\n" + "="*80)
        print("VARIANTES CLINICAMENTE RELEVANTES:")
        print("="*80)
        for idx, row in cancer_variants.iterrows():
            print(f"\nGene: {row.get('SYMBOL', 'N/A')}")
            print(f"Localização: {row.get('Location', 'N/A')}")
            print(f"Consequência: {row.get('Consequence', 'N/A')}")
            print(f"HGVSc: {row.get('HGVSc', 'N/A')}")
            print(f"HGVSp: {row.get('HGVSp', 'N/A')}")
            print(f"Impacto: {row.get('IMPACT', 'N/A')}")
            print(f"Frequência alélica (1000 Genomes): {row.get('AF', 'N/A')}")
            print("-"*80)
    else:
        print("Nenhuma variante encontrada nos genes de câncer de pulmão conhecidos.")
        print("Verificando todas as variantes...")
else:
    print("Coluna SYMBOL não encontrada. Exibindo todas as variantes:")

Variantes encontradas em genes conhecidos de câncer de pulmão: 2

Genes identificados:
SYMBOL
EGFR    1
BRAF    1
Name: count, dtype: int64

VARIANTES CLINICAMENTE RELEVANTES:

Gene: EGFR
Localização: NC_000007.14:55191822
Consequência: missense_variant
HGVSc: ENST00000275493.7:c.2573T>G
HGVSp: ENSP00000275493.2:p.Leu858Arg
Impacto: MODERATE
Frequência alélica (1000 Genomes): -
--------------------------------------------------------------------------------

Gene: BRAF
Localização: NC_000007.14:140753336
Consequência: missense_variant
HGVSc: ENST00000646891.2:c.1799T>A
HGVSp: ENSP00000493543.1:p.Val600Glu
Impacto: MODERATE
Frequência alélica (1000 Genomes): -
--------------------------------------------------------------------------------


## Interpretacao dos Resultados


### Gene Identificado:
- **EGFR**
- **BRAF**

### Variante Identificada:
- **EGFR**
    - **Nomenclatura HGVSc:** ENST00000275493.7:c.2573T>G
    - **Nomenclatura HGVSp:** ENSP00000275493.2:p.Leu858Arg
    - **Tipo de variante:** missense
    - **Localização:** NC_000007.14:55191822
    - **Impacto:** MODERATE

- **BRAF**
    - **Nomenclatura HGVSc:** ENST00000646891.2:c.1799T>A
    - **Nomenclatura HGVSp:** ENSP00000493543.1:p.Val600Glu
    - **Tipo de variante:** missense
    - **Localização:** NC_000007.14:140753336
    - **Impacto:** MODERATE

### Significância Clínica:
- **EGFR**
    Variacao eh oncogenica, ocorrendo no Exon 21 do dominio da tirosina quinase. Deixa a TK constantemente ativada, independendo de vias de sinalizacao celular. Comum em casos de canceres de pulmao.
    [Referencia](https://www.oncokb.org/gene/EGFR/L858R?refGenome=GRCh38)

- **BRAF**
    - Variacao eh oncogenica, ocorrendo no Exon 15 do dominio kinase, comum em casos de melanoma, e canceres de pulmao ou tireoide.
    - Mutacao deixa a via BRAF constantemente ativada na forma monomerica, mas ainda sensiveis a inibidores RAF.
    [Referencia](https://www.oncokb.org/gene/BRAF/V600E)



### Conduta Clínica Sugerida:
- **EGFR**
    - Tratamento com inibidores de tirosina quinase de 1a,2a, ou 3a geracao
    
- **BRAF**
    - Tratamento com inibidores de tirosina quinase, ou inibidores RAF

### Opções Terapêuticas:
- Tratamento de dupla acao, inibindo as TKs, junto do dominio MAPK do paciente
    - Dabrafenib + Trametinib
